ca/ocsp_test.go

```diff
@@ -0,0 +1,133 @@
+package ca
+
+import (
+	"encoding/hex"
+	"testing"
+	"time"
+
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/metrics"
+	"github.com/letsencrypt/boulder/test"
+	"golang.org/x/crypto/ocsp"
+)
+
+func serial(t *testing.T) []byte {
+	serial, err := hex.DecodeString("aabbccddeeffaabbccddeeff000102030405")
+	if err != nil {
+		t.Fatal(err)
+	}
+	return serial
+
+}
+
+// Set up an ocspLogQueue with a very long period and a large maxLen,
+// to ensure any buffered entries get flushed on `.stop()`.
+func TestOcspLogFlushOnExit(t *testing.T) {
+	t.Parallel()
+	log := blog.NewMock()
+	stats := metrics.NoopRegisterer
+	queue := newOCSPLogQueue(4000, 10000*time.Millisecond, stats, log)
+	go queue.loop()
+	queue.enqueue(serial(t), time.Now(), ocsp.ResponseStatus(ocsp.Good))
+	queue.stop()
+
+	expected := []string{
+		"INFO: [AUDIT] OCSP signed: aabbccddeeffaabbccddeeff000102030405:0,",
+	}
+	test.AssertDeepEquals(t, log.GetAll(), expected)
+}
+
+// Ensure log lines are sent when they exceed maxLen.
+func TestOcspFlushOnLength(t *testing.T) {
+	t.Parallel()
+	log := blog.NewMock()
+	stats := metrics.NoopRegisterer
+	queue := newOCSPLogQueue(100, 100*time.Millisecond, stats, log)
+	go queue.loop()
+	for i := 0; i < 5; i++ {
+		queue.enqueue(serial(t), time.Now(), ocsp.ResponseStatus(ocsp.Good))
+	}
+	queue.stop()
+
+	expected := []string{
+		"INFO: [AUDIT] OCSP signed: aabbccddeeffaabbccddeeff000102030405:0,aabbccddeeffaabbccddeeff000102030405:0,",
+		"INFO: [AUDIT] OCSP signed: aabbccddeeffaabbccddeeff000102030405:0,aabbccddeeffaabbccddeeff000102030405:0,",
+		"INFO: [AUDIT] OCSP signed: aabbccddeeffaabbccddeeff000102030405:0,",
+	}
+	test.AssertDeepEquals(t, log.GetAll(), expected)
+}
+
+// Ensure log lines are sent after a timeout.
+func TestOcspFlushOnTimeout(t *testing.T) {
+	t.Parallel()
+	log := blog.NewMock()
+	stats := metrics.NoopRegisterer
+	queue := newOCSPLogQueue(90000, 10*time.Millisecond, stats, log)
+
+	go queue.loop()
+	for i := 0; i < 3; i++ {
+		queue.enqueue(serial(t), time.Now(), ocsp.ResponseStatus(ocsp.Good))
+		queue.enqueue(serial(t), time.Now(), ocsp.ResponseStatus(ocsp.Good))
+		// This gets a little tricky: Each iteration of the `select` in
+		// queue.loop() is a race between the channel that receives log
+		// events and the `<-clk.After(n)` timer. Even if we used
+		// a fake clock, our loop here would often win that race, producing
+		// inconsistent logging results. For instance, it would be entirely
+		// possible for all of these `enqueues` to win the race, putting
+		// all log entries on one line.
+		// To avoid that, sleep using the wall clock for 50ms.
+		time.Sleep(50 * time.Millisecond)
```

Comment:
Eep, sleeps in tests are no fun. They are often sources of flakiness, and make tests noticeably slow (this one alone takes .15 seconds, or 9 frames!).

How about instead: use a fake clock, enqueue just one small set of serials, move the fake clock forward, and then poll the contents of log.GetAll() a set number of times (maybe 10?). If it never shows up in those polls, then fail the test.

Yes, this still has a chance of flakiness, but will be much faster. Would be even better for this test to be able to listen on a notification channel, but I don't think either blog nor the prometheus metric offers that. Other ideas?

Reply:
I tried this, and it doesn't work; `log.GetAll()` never becomes nonempty. I'm not sure exactly why; it seems quite possible there's a resource starvation issue (i.e. the test's goroutine never yields), or that the fake clock isn't working quite how I expect.

Reply:
I dug into this a little more. There are actually two problems with the fake clock approach:

`fake.Sleep(...)` advances the fake clock's internal time, and fires any timers that have been set. However, there's a race between `clk.After(..)` in `go queue.loop()` and `fake.Sleep(...)` in the test. If `fake.Sleep(...)` runs before `clk.After(..)`, the resulting timer will be configured to fire after some future sleep, which never happens. I can't think of a way to resolve this, short of adding some synchronization code in `loop()` that only runs for tests and notifies the test that it's okay to proceed. That seems to break the realness of the test and make the code more complex.

The second issue is that the log lines are getting written from a goroutine separate from the test goroutine, and so the test isn't guaranteed to see the log lines after it sends them off. This could be solved by adding e.g. a channel to blog.Mock, and allowing consumer to block on it. But I think that's outside the scope of this change.

I do agree that putting sleeps in tests sucks, and I would really like to find a better way, but I can't see one here.

Reply:
Alright, that makes sense. Thanks for digging in to it!

(Also, boo to github, which has decided that this comment thread no longer needs to show up in the diff itself, only showing it in the conversation page.)